src/utils/timer.ts

```diff
@@ -0,0 +1,93 @@
+import { logWarn } from './logger';
+
+interface TimerProps {
+    /** Set a upper threshold for the timer, which will make the timer autostop. Default is 2000 milliseconds. */
+    maxTime?: number;
+    /** Optional callback. Check the this reference binding as it will default to {Timer} */
+    callback?: () => void;
+}
+
+type TimerStatus = 'Started' | 'Stopped' | 'Over time' | 'Not started';
+
+export class Timer {
+    /**
+     * Once timer stops, the time elapsed is stored here.
+     * For simplicity, this prop is kept private and consumer should use timeElapsed instead.
+     */
+    private _finalTime = 0;
+
+    private _startTime?: Date;
+
+    /** An optional callback when the timer stops. */
+    private _callback?: () => void;
+
+    /**
+     * We use a max time for the timer to avoid it counting "forever".
+     * This value is configurable during construction.
+     */
+    private _maxTime: number;
+    private _timerId?: NodeJS.Timeout;
+    private _status: TimerStatus;
+
+    constructor(props?: TimerProps) {
+        if (props && props.maxTime && typeof props.maxTime === 'number' && !isNaN(props.maxTime)) {
+            this._maxTime = props.maxTime;
+        } else this._maxTime = 2000;
+
+        this._status = 'Not started';
+        this._callback = props?.callback;
+    }
+
+    public start(): void {
+        if (!this._timerId) {
+            this._startTime = new Date();
+            this._status = 'Started';
+            this._timerId = globalThis.setTimeout(() => {
+                this._finalTime = this.stop();
+                logWarn('A timer has exceeded its max time at ', this._finalTime);
+                this._status = 'Over time';
+            }, this._maxTime);
+        } else {
+            throw new Error('A timer is already active.');
```

Comment:
I'm wondering if we should omit this throw, and just allow to call start several times? Where it doesn't do anything if it's already started. This is how the timer behaves in c#. If se use start in a uiComponent or in useEffect it might get called multiple times.
In c#, in addition to start, there is a restart which will just reset the startTime, but not sure if this is needed. When there is a restart method, it makes it obvious that start will only start it the first time it's called, and continue counting if it's called multiple times.